REGRESSION (286777@main): [iOS] Crash when tapping on date input on devices with 12-hour time

pxlcoder opened the pull request:
#### 8af24b25e9e2fec43f81fd4d5f9d3c0b275f5b04
<pre>
REGRESSION (286777@main): [iOS] Crash when tapping on date input on devices with 12-hour time
<a href="https://bugs.webkit.org/show_bug.cgi?id=285106">https://bugs.webkit.org/show_bug.cgi?id=285106</a>
<a href="https://rdar.apple.com/141879717">rdar://141879717</a>

Reviewed by Richard Robinson.

286777@main introduced a new use of `NSDateFormatter` in the date picker
initialization path. This formatter did not specify a locale, but used a
format string with 24-hour time.

When working with fixed format dates, the `locale` property should be a POSIX
locale, and the `timeZone` property should be UTC, as described in
<a href="https://developer.apple.com/documentation/foundation/nsdateformatter.">https://developer.apple.com/documentation/foundation/nsdateformatter.</a>

Without specifying the aforementioned properties, `dateFromString:` may return
`nil`. A crash would then be encountered when creating an `NSDateInterval` with
a `nil` end date.

No new tests. This issue is reproducible when setting `AppleICUForce12HourTime`
to `YES` in the global domain. However, the preference must be set prior to
process launch, which is not possible when writing an API test. Additionally,
it is not desirable to set this bit for all tests.

* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):

Canonical link: <a href="https://commits.webkit.org/288258@main">https://commits.webkit.org/288258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e3cbb9b7ad1e34f3119238e6a3d442275090bfc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82258 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87390 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33319 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64111 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21858 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85327 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44389 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1291 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32360 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29651 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88746 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6826 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72503 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9790 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70662 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71721 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15857 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14872 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/919 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12769 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9517 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15038 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9391 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12857 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11164 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -87,6 +87,8 @@ - (id)initWithView:(WKContentView *)view inputType:(WebKit::InputType)inputType
+    [maximumDateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
     [maximumDateFormatter setDateFormat:kDateTimeFormatString];
     RetainPtr maximumDate = [maximumDateFormatter dateFromString:@"10000-12-31T23:59"]; // UIDatePicker cannot have more than 10,000 selectable years
     _dateInterval = adoptNS([[NSDateInterval alloc] initWithStartDate:[NSDate distantPast] endDate:maximumDate.get()]);
```
should there be a RELEASE_ASSERT added here to make sure maximumDate is never nil?

pxlcoder replied:
There's already one in Foundation, so not sure it's really necessary?

pxlcoder replied:
Also the existing crash log is very clear: "endDate is nil".

pxlcoder committed:
REGRESSION (286777@main): [iOS] Crash when tapping on date input on devices with 12-hour time
https://bugs.webkit.org/show_bug.cgi?id=285106
rdar://141879717

Reviewed by Richard Robinson.

286777@main introduced a new use of `NSDateFormatter` in the date picker
initialization path. This formatter did not specify a locale, but used a
format string with 24-hour time.

When working with fixed format dates, the `locale` property should be a POSIX
locale, and the `timeZone` property should be UTC, as described in
https://developer.apple.com/documentation/foundation/nsdateformatter.

Without specifying the aforementioned properties, `dateFromString:` may return
`nil`. A crash would then be encountered when creating an `NSDateInterval` with
a `nil` end date.

No new tests. This issue is reproducible when setting `AppleICUForce12HourTime`
to `YES` in the global domain. However, the preference must be set prior to
process launch, which is not possible when writing an API test. Additionally,
it is not desirable to set this bit for all tests.

* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):

Canonical link: https://commits.webkit.org/288258@main
